client/block.c

```diff
@@ -1992,15 +1985,95 @@ void xdag_block_finish()
 	pthread_mutex_lock(&block_mutex);
 }
 
-int xdag_get_block_info(xdag_hash_t hash, void *data, int (*callback)(void*, int, xdag_hash_t, xdag_amount_t, xdag_time_t, xdag_remark_t))
+int xdag_get_block_info(xdag_hash_t hash, void *data, int (*callback)(void*, int, xdag_hash_t, xdag_amount_t, xdag_time_t, const xdag_remark_t))
 {
 	pthread_mutex_lock(&block_mutex);
 	struct block_internal *bi = block_by_hash(hash);
 	pthread_mutex_unlock(&block_mutex);
 
 	if(callback) {
-		return callback(data, bi->flags & ~BI_OURS,  bi->hash, bi->amount, bi->time, bi->remark ? bi->remark : "");
+		return callback(data, bi->flags & ~BI_OURS,  bi->hash, bi->amount, bi->time, ((bi->flags & BI_REMARK) && bi->bremark != NULL) ? (const char*)bi->bremark->remark : "" );
 	}
 
 	return 0;
 }
+
+size_t remark_acceptance(struct block_internal* bi, xdag_remark_t origin, xdag_remark_t strbuf)
+{
+	size_t size = strnlen(origin, sizeof(xdag_remark_t));
+	if(size == 0){
+		xdag_debug("String is empty. [function remark_acceptance]");
+		bi->flags &= ~BI_REMARK;
+		*strbuf = '\0';
+		return 0;
+	}
+	if(sizeof(xdag_remark_t) == size){
+		xdag_debug("Remark is not a string. [function remark_acceptance]");
+		bi->flags &= ~BI_REMARK;
+		*strbuf = '\0';
+		return 0;
+	}
+	if(!validate_ascii(origin)){
+		xdag_debug("Remark is not a valid ascii string. [function remark_acceptance]");
+		bi->flags &= ~BI_REMARK;
+		*strbuf = '\0';
+		return 0;
+	}
+	++size;
+	memcpy(strbuf, origin, size);
+	return size;
+}
+
+void add_remark_bi(struct block_internal* bi, xdag_remark_t strbuf, size_t size)
+{
+	if(bi->bremark != NULL) {
+		xdag_fatal("Unexpected error, bremark already exist [function add_remark_bi]");
+		bi->flags &= ~BI_REMARK;
+		return;
+	}
+	bi->bremark = xdag_malloc(sizeof(struct block_remark) + size);
+	if(bi->bremark == NULL) {
+		xdag_err("xdag_malloc failed, [function add_remark_bi]");
+		bi->flags &= ~BI_REMARK;
+		return;
+	} else {
+		memset(bi->bremark, 0, sizeof(struct block_remark));
+		memcpy(bi->bremark->remark, strbuf, size);
+	}
+}
+
+void add_backref(struct block_internal* blockRef, struct block_internal* nodeBlock)
+{
+	int i = 0;
+	struct block_backrefs **blockRefs_ptr;
+	if(blockRef->flags & BI_REMARK){
```

Comment:
backrefs is a LIFO list.
if the block has the remark, the first element of the list is ALWAYS the remark.
successive element of the list are backrefs